Native windows port of the telepresence CLI

thallgren opened the pull request:
## Description

This PR makes it possible to compile and run the telepresence CLI on Windows. This build system is not yet ported so tests that require that will not run on Windows yet.

In order to build and run, you will need to do the following (I didn't put this in DEVELOPING.md because I think it will be very short lived):
1. Determine a telepresence version, e.g. `v2.3.4` and ensure that the cluster that you want to connect to have access to a traffic-manager and traffic agent of that version from a docker repository.
2. Build or install the windows binary using `go build` or `go install` with an included `-ldflags` to inject the chosen version:
```
go install -trimpath -ldflags='-X=github.com/telepresenceio/telepresence/v2/pkg/version.Version=v2.3.4
```
3. Ensure that the `wintun.dll` downloaded from https://www.wintun.net/ is available to the dll loader (for instance, put it adjacent to the telepresence binary).
4. Enjoy ;-)

## Checklist

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Unset GOOS and GOARCH before building tools.

Cross-compiling is very convenient when working with changes that affect
more than one platform. But cross-compiling the tools is not something
that is ever desired.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Change all `unix.IPPROTO_XXX` to `ipproto.XXX`

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move safeCobraCommand interface into cliutil package and export it.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Create generic Run, Background, and BackgroundAsRoot functions.

Breaking part of start of connector and daemon out to generic methods
because they look very different on Windows.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move client/util_exec.go to cliutil/exec.go

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Adopt socket logic to Windows

Split the socket logic into unix and windows files and have the windows
sockets use named pipes instead. Unix sockets should work on Windows in
theory but they caused a lot of problems with dangling sockets and
permission problems. The named pipes however, work just fine.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Adopt rotating log to windows.

This commit is mainly about handling the extended variants of the
info struct that is returned from `os.Stat()` and changing owner and
group on the log files. Windows also has some weirdness around moving
opened files that motivated the os-specific functions for opening the
log files.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure close of both command and children when using dpipe.

The `sshfs` process starts child processes, and lo and behold, such
processes are not killed when killing a parent process on Windows. Not
unless special tricks are made. This commit adds those tricks.

On Windows, `DPipe()` will use `TerminateProcess()` to terminate the
processes recursively, children first.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Adopt sshfs volume mounts to windows.

This implementation relies on that `win-sshfs` is installed on the
workstation. The one used for verification was downloaded from:
https://github.com/billziss-gh/sshfs-win/releases/tag/v3.7.21011

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Pass context down to openTun() function.

This is in preparation for adding the Windows TUN-device.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add Windows specific TUN device based on Wireguard's wintun.

The device requires that `wintun.dll` is available to the DLL loader. It
can be downloaded from https://www.wintun.net/

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Use datawire/dtest instead of datawire/ambassador/pkg/dtest

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on pkg/client/cli/cliutil/exec.go:
```diff
@@ -3,75 +3,61 @@ package cliutil
 	s, err := cmd.Process.Wait()
 	if err != nil {
-		return fmt.Errorf("%s: %w", logging.ShellString(exe, args), err)
+		return fmt.Errorf("%s: %v", logging.ShellString(exe, args), err)
```
hmm why not `%w`? what kind of errors can we get from cmd.Process.Wait()?

josecv commented on pkg/client/cli/cliutil/exec_windows.go:
```diff
@@ -0,0 +1,44 @@
+
+func shellExec(_ context.Context, verb, exe string, args []string) error {
+	cwd, _ := os.Getwd()
+	verbPtr, _ := windows.UTF16PtrFromString(verb)
+	exePtr, _ := windows.UTF16PtrFromString(exe)
+	cwdPtr, _ := windows.UTF16PtrFromString(cwd)
```
It looks like these can only error if the string contains a NUL byte somewhere, so it makes sense to ignore the error -- should we add a short comment explaining this?

josecv commented on pkg/client/sockets_windows.go:
```diff
@@ -0,0 +1,64 @@
+	ConnectorSocketName = `\\.\pipe\telepresence-connector`
+
+	// DaemonSocketName is the path used when communicating to the daemon process
+	DaemonSocketName = `\\.\pipe\telepresence-daemon`
```
it seems like these aren't paths so much as just pipe names, is that right? If so maybe we should update the comment to reflect that, and perhaps add a link to [the windows docs](https://docs.microsoft.com/en-us/windows/win32/ipc/pipe-names)

josecv commented on pkg/client/daemon/service.go:
```diff
@@ -190,19 +188,10 @@ func run(c context.Context, loggingDir, configDir, dns string) error {
-		origUmask := unix.Umask(0)
-		grpcListener, err = net.Listen("unix", client.DaemonSocketName)
-		unix.Umask(origUmask)
+		grpcListener, err = client.ListenSocket(c, processName, client.DaemonSocketName)
```
how does the cleanup work on windows? Will [these lines](https://github.com/telepresenceio/telepresence/pull/1864/commits/1cb5724deccca2c75c9c07e5ed6f2b53c2e49edd#diff-f541fcdb8230f0df59b1ecefcedb4056040fb4d4e68b15acb6906daaeef8e575R152-R154) actually clean anything up when using a named pipe?

thallgren replied:
The "these lines" link doesn't make much sense to me.

thallgren replied:
Oh, you're referring to:
```
	defer func() {
		if grpcListener != nil {
			_ = os.Remove(grpcListener.Addr().String())
		}
	}()
```
No, that won't do on Windows.

josecv commented on pkg/client/sockets_windows.go:
```diff
@@ -0,0 +1,64 @@
+	return conn, err
+}
+
+const AllowEveryone = "S:(ML;;NW;;;LW)D:(A;;0x12019f;;;WD)"
```
This is... something else -- is there a reference doc or something where we can decode this magic string?

thallgren replied:
But ... isn't that self evident? ;-)

josecv commented on pkg/client/sockets_windows.go:
```diff
@@ -0,0 +1,64 @@
+	if err != nil {
+		return false
+	}
+	h, err := windows.CreateFile(uPath, windows.GENERIC_READ|windows.GENERIC_WRITE, 0, nil, windows.OPEN_EXISTING, windows.FILE_FLAG_OVERLAPPED, 0)
```
this system call is very badly named -- could we add a comment stating that the `OPEN_EXISTING` disposition will make `CreateFile` fail if the file doesn't already exist? At first glance it looks like "Create a file or open it if it's already there"

josecv commented on pkg/client/logging/sysinfo.go:
```diff
@@ -2,18 +2,18 @@ package logging
-type sysinfo interface {
-	birthtime() time.Time
+type SysInfo interface {
+	Birthtime() time.Time
```
nit: should this be called `BirthTime`?

thallgren replied:
It should. Compound words in English bugs me. You use "namespace" as one word but "birth time" as two. Weird. In Swedish, we make it simple. Always concatenate :-)

thallgren replied:
Or "timestamp" versus "time span". What?!

josecv commented on pkg/client/logging/sysinfo_windows.go:
```diff
@@ -0,0 +1,98 @@
+		&wi.group,
+		&wi.dacl,
+		&wi.sacl,
+		&wi.secDesc,
```
the [internet says](https://docs.microsoft.com/en-us/windows/win32/api/aclapi/nf-aclapi-getnamedsecurityinfoa) that this buffer needs to be freed by the user -- does that apply here? do we need to do something special to garbage collect `secDesc`?

thallgren replied:
Good catch!

josecv commented on pkg/client/logging/sysinfo_windows.go:
```diff
@@ -0,0 +1,98 @@
+		&wi.secDesc,
+	)
+	if err != nil {
+		panic(err)
```
i don't know how I should feel about this -- should we at least rename this function to `MustGetSysInfo` to denote that it can panic? What kinds of errors can we get out of `GetNamedSecurityInfo`?

josecv commented on pkg/dpipe/dpipe_windows.go:
```diff
@@ -0,0 +1,25 @@
+	<-ctx.Done()
+
+	// A process is sometimes not terminated gracefully by the SIGTERM, so we give
+	// it a second to succeed and then kill it forcefully.
```
these comments don't seem to apply here

josecv commented on pkg/client/daemon/dns/flush_windows.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+func Flush(c context.Context) {
+	// TODO
```
should this be implemented? I guess `setDNS` actually flushes, so maybe this comment should just be updated to explain why this isn't implemented?

thallgren replied:
I don't think `setDNS()` performs a complete flush. It will just flush the device. Windows may have other caches in other places and we should find out where and implement a proper flush here. I'll update the TODO comment.

thallgren replied:
Doh. You're right. The `setDNS()` calls `ipconfig /flushdns`.

josecv reviewed (commented):
A lot of new code, and I don't really feel qualified to pass judgement on the big picture, but it all seems to make sense to me.

Mostly just have questions and suggestions for comments; my biggest concern is around cleaning up the sockets, as I'm not sure if it's possible to simply `os.Remove` a named pipe

Commit:
Changes in response to Jose's PR review

- Use %w when wrapping errors
- Add comment about ignoring errors from `UTF16PtrFromString()`
- Don't remove socket based on listener address on Windows
- Replace TODO comment
- Have GetSysInfo() return errors instead of panicking
- Birthtime -> BirthTime
- Free security descriptor returned by `GetNamedSecurityInfo()`
- Remove incorrect comment about termination timer

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented:
@josecv I think I've addressed all you comments now. I also rebased the PR on top of the current release/v2.

Commit:
Info on log file that cannot be deleted (remove when resolved)

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Bumbing to go 1.16 because 1.15 doesn't have os.ReadDir()

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that go is properly installed before running lint.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't build splitToUDPAddr on windows. It's not used.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Euthpic commented:
I am puzzled about this step(newbie to go):
```
Build or install the windows binary using go build or go install with an included -ldflags to inject the chosen version:
go install -trimpath -ldflags='-X=github.com/telepresenceio/telepresence/v2/pkg/version.Version=v2.3.4
```
I installed go sdk and set all the GOPATH/GOBIN/GOROOT,but this command still doesn't work correctly:
```
package .: no Go files in D:\download
```
Then I download the realease from https://github.com/telepresenceio/telepresence/archive/refs/tags/v2.3.4.zip.
And run this command in /d/download/telepresence-2.3.4/cmd/telepresence,it still faied:
```
# github.com/telepresenceio/telepresence/v2/pkg/dpipe
..\..\pkg\dpipe\dpipe.go:46:27: undefined: unix.SIGKILL
..\..\pkg\dpipe\dpipe.go:50:26: undefined: unix.SIGTERM
# github.com/telepresenceio/telepresence/v2/pkg/tun
..\..\pkg\tun\device.go:14:18: undefined: Device
..\..\pkg\tun\device.go:15:9: undefined: openTun
..\..\pkg\tun\device.go:20:10: undefined: Device
..\..\pkg\tun\device.go:26:10: undefined: Device
..\..\pkg\tun\device.go:31:10: undefined: Device
..\..\pkg\tun\device.go:37:10: undefined: Device
..\..\pkg\tun\device.go:43:10: undefined: Device
..\..\pkg\tun\device.go:47:10: undefined: Device
..\..\pkg\tun\device.go:52:13: undefined: unix.Socket
..\..\pkg\tun\device.go:52:33: undefined: unix.SOCK_DGRAM
..\..\pkg\tun\device.go:52:33: too many errors
# github.com/telepresenceio/telepresence/v2/pkg/connpool
..\..\pkg\connpool\connid.go:58:22: undefined: unix.IPPROTO_TCP
..\..\pkg\connpool\connid.go:83:22: undefined: unix.IPPROTO_TCP
..\..\pkg\connpool\connid.go:106:7: undefined: unix.IPPROTO_TCP
..\..\pkg\connpool\connid.go:112:7: undefined: unix.IPPROTO_UDP
..\..\pkg\connpool\connid.go:137:10: undefined: unix.IPPROTO_TCP
..\..\pkg\connpool\connid.go:139:10: undefined: unix.IPPROTO_UDP
..\..\pkg\connpool\connid.go:141:10: undefined: unix.IPPROTO_ICMP
..\..\pkg\connpool\connid.go:143:10: undefined: unix.IPPROTO_ICMPV6
..\..\pkg\connpool\connid.go:151:7: undefined: unix.IPPROTO_ICMP
..\..\pkg\connpool\connid.go:153:7: undefined: unix.IPPROTO_TCP
..\..\pkg\connpool\connid.go:153:7: too many errors
# github.com/telepresenceio/telepresence/v2/pkg/client/logging
..\..\pkg\client\logging\rotatingfile.go:146:17: undefined: getSysInfo
..\..\pkg\client\logging\rotatingfile.go:208:10: undefined: dupToStd
..\..\pkg\client\logging\rotatingfile.go:246:15: undefined: getSysInfo
..\..\pkg\client\logging\rotatingfile.go:261:38: undefined: getSysInfo
# github.com/telepresenceio/telepresence/v2/pkg/tun/ip
..\..\pkg\tun\ip\header.go:128:26: undefined: unix.IPPROTO_UDP
```

Which step was wrong?

LukeShu commented on pkg/ipproto/ipproto_windows.go:
```diff
@@ -0,0 +1,10 @@
+	UDP    = syscall.IPPROTO_UDP
+	ICMP   = 0x1
+	ICMPV6 = 0x3a
+)
```
The numbers are the same on all platforms (by definition)... if we can't just grab them all from the same syscall-ish package, I'd just write the raw numbers and throw in a comment pointing at https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml

LukeShu commented on pkg/client/cli/cliutil/exec.go:
```diff
@@ -0,0 +1,63 @@
+	cmd.Stderr = sCmd.ErrOrStderr()
+	cmd.Stdin = sCmd.InOrStdin()
+	for k, v := range env {
+		cmd.Env = append(os.Environ(), k+"="+v)
```
Isn't this going to discard all but the last item in `env`?

Shouldn't this be
```go
cmd.Env = os.Environ()
for k, v := range env {
	cmd.Env = append(cmd.Env, k+"="+v)
```

LukeShu commented on pkg/client/cli/cliutil/exec.go:
```diff
@@ -0,0 +1,63 @@
+	cmd := exec.CommandContext(ctx, exe, args...)
+	cmd.Stdout = sCmd.OutOrStdout()
+	cmd.Stderr = sCmd.ErrOrStderr()
+	cmd.Stdin = sCmd.InOrStdin()
```
Here's something that slipped in recently that you may not have noticed: The tests don't do `cmd.SetOut` or anything anymore; they fork the actual process.  So it shouldn't be necessary to explicitly pass around stdio anymore (whether via a cobra command or via separate stdin/out/err arguments).

josecv replied:
So removing these lines actually resulted in [a test failure](https://app.circleci.com/pipelines/github/telepresenceio/telepresence/2245/workflows/b4878a93-5976-4110-8e9a-f1c210f19aff/jobs/11744?invite=true#step-108-240) :(

LukeShu replied:
Not removing them, but just using `cmd.Stdout = os.Stdout` instead; no need to take the `sCmd` argument.

josecv replied:
OOH okay my bad

LukeShu replied:
Also note that if we don't need to pass in `sCmd`, then there's no reason to move SafeCobraCommand in to cliutil.

LukeShu commented on pkg/client/cli/cliutil/exec_unix.go:
```diff
@@ -0,0 +1,74 @@
+func background(ctx context.Context, exe string, args []string) error {
+	// The context should not kill it if cancelled
+	ctx = dcontext.WithoutCancel(ctx)
+	cmd := exec.CommandContext(ctx, exe, args...)
```
If you're using stdlib `exec` instead of dlib `dexec`, the *only* thing that the ctx given to `CommanContext` does is cancel it, so if you're using `WithoutCancel`, might as well just use `dexec.Command` (no Context).

LukeShu commented on pkg/client/cli/cliutil/exec_unix.go:
```diff
@@ -0,0 +1,74 @@
+		args = append([]string{"--non-interactive", "--preserve-env", exe}, args...)
+		exe = "sudo"
+	}
+	return Background(ctx, exe, args)
```
I'm not a fan of this business with having `exe` be separate than just `args[0]`.

josecv replied:
This is the only comment I didn't really address for two reasons:

1. I think it's pretty reasonable, especially in code like this where the executable is being modified. IMHO it'd look a little bit weird to do `args[0] = "sudo"` instead of `exe = "sudo"`
2. This is consistent with the `dexec.Command*` APIs

LukeShu commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -53,53 +46,7 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
-	}
-
-	return nil
+	return BackgroundAsRoot(ctx, client.GetExe(), []string{"daemon-foreground", logDir, configDir, dnsIP})
```
I'm already not a fan of having the `exe` be separate from `args[0]`, but declaring it as `args []string` instead of `args ...string` just makes this clunky when it could be `return BackgroundAsRoot(ctx, client.GetExe(), "daemon-foreground", logDir, configDir, dnsIP)`.

LukeShu commented on pkg/client/cli/cliutil/exec.go:
```diff
@@ -0,0 +1,63 @@
+
+	// Ensure that interrupt is propagated to the child process
+	sigCh := signalNotifications()
+	defer close(sigCh)
```
This is panicy.  The `os/signal.Notify` machinery will panic if it tries to write to this closed signal.  Need to call `signal.Stop(sigCh)`.

josecv commented:
@Euthpic we'll be releasing a Developer Preview of the windows client this week 🙂 I suggest waiting until we do that so you don't have to go through the motions of building it yourself

josecv committed:
Incorporate PR Feedback

Signed-off-by: Jose Cortes <josecortes@datawire.io>

LukeShu commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -53,53 +46,7 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
-				return err
-			}
-		}
-		args = append([]string{"sudo", "--non-interactive", "--preserve-env"}, args...)
```
@josecv wrote
>  IMHO it'd look a little bit weird to do args[0] = "sudo" instead of exe = "sudo"

I don't think this looked weird.  I'd argue that it, in fact, read much easier.

LukeShu commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -53,53 +46,7 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
-		// depending on `passprompt_override` in `/etc/sudoers`.  So we'll do a pre-flight
-		// `sudo --non-interactive true` to decide whether to display it.
-		//
-		// Note: Using `sudo --non-interactive --validate` does not work well in situations
-		// where the user has configured `myuser ALL=(ALL:ALL) NOPASSWD: ALL` in the sudoers
-		// file. Hence the use of `sudo --non-interactive true`. A plausible cause can be
-		// found in the first comment here:
-		// https://unix.stackexchange.com/questions/50584/why-sudo-timestamp-is-not-updated-when-nopasswd-is-set
```
This comment went away... for no real reason?  IMO it was a very helpful comment.

LukeShu commented on pkg/client/cli/cliutil/daemon.go:
```diff
@@ -53,53 +46,7 @@ func launchDaemon(ctx context.Context, dnsIP string) error {
-		if err := needPwCmd.Run(); err != nil {
-			fmt.Printf("Need root privileges to run: %s\n", logging.ShellString(args[0], args[1:]))
-			// `sudo` won't be able to read the password from the terminal when we run
-			// it with Setpgid=true, so do a pre-flight `sudo --validate` to read the
-			// password, and then enforce that being re-used by passing
-			// `--non-interactive`.
-			pwCmd := dexec.CommandContext(ctx, "sudo", "--validate")
```
These 2 `--validate`s were changed to `true`s.  Was that intentional?  There was a lot shoved in to that one commit.

LukeShu commented on pkg/client/cli/cliutil/exec.go:
```diff
@@ -0,0 +1,69 @@
+	return nil
+}
+
+// Background will start the given exe as a detach process.
+func Background(ctx context.Context, exe string, args ...string) error {
+	return background(ctx, exe, args...)
+}
+
+// BackgroundAsRoot will start the given exe as a detach process with elevated privileges.
+func BackgroundAsRoot(ctx context.Context, exe string, args ...string) error {
+	return backgroundAsRoot(ctx, exe, args...)
+}
```
These aren't great function names... where's the verb?

LukeShu commented on pkg/client/cli/cliutil/exec_unix.go:
```diff
@@ -0,0 +1,72 @@
+	return Background(ctx, exe, args...)
+}
+
+func signalNotifications() chan os.Signal {
+	// Ensure that SIGINT and SIGTERM are propagated to the child process
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, unix.SIGINT, unix.SIGTERM)
+	return sigCh
+}
```
Why repeat this business several times, instead of just having each platform declare a `signalsToForward` variable?

LukeShu commented:
I didn't look at the `dpipe` changes at all, since I know it's getting rewritten in @josecv's PR.

LukeShu commented on .circleci/config.yml:
```diff
@@ -26,7 +26,7 @@ commands:
       version:
         type: string
-        default: "1.15"
+        default: "1.16"
```
The only thing that actually ended up using Go 1.16 was one place ware it said `io.Discard` instead of the older `ioutil.Discard`.

LukeShu commented:
Having the `Background` and `BackgroundAsRoot` functions in `cliutil` just seems a little wonky to me.  I'd be inclined to put them in `proc` instead.

LukeShu commented on go.mod:
```diff
@@ -18,11 +21,13 @@ require (
 	github.com/spf13/cobra v1.1.1
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.0
-	github.com/telepresenceio/telepresence/rpc/v2 v2.3.7
-	golang.org/x/net v0.0.0-20210226172049-e18ecbb05110
+	github.com/telepresenceio/telepresence/rpc/v2 v2.3.6
```
*Downgrading* `rpc/v2` is a little weird... I suspect it's the artifact of a bad rebase?
